AP_Scripting/AP_Vehicle: add set_target_position for use with Copter and Rover (at least)

rmackay9 opened the pull request:
This PR allows a script to request a Copter or Rover fly/drive to a specified location in Guided mode.

The more detailed changes are:

- AP_Vehicle gets a new set_target_position virtual method
- Copter and Rover override the above method and then call the existing guided mode methods
- AP_Scripting is given bindings for the new function and Location's alt field is also exposed
- A new example script, set-target-location.lua is added and has been tested succesfully on Copter.  It also works on Rover if the copter_guided_mode_num is set to "15" and copter_land_mode_num is set to "4" (for Hold mode)

This is a first step for adding the ability to control the vehicle's movement from a script, future steps probably include:

- add takeoff(target_alt)
- add reached_destination()?
- implement for Plane and Sub?
- add velocity controller for Copter/Rover

WickedShell commented on libraries/AP_Scripting/generator/description/bindings.desc:
```diff
@@ -4,6 +4,7 @@ include AP_Common/Location.h
 
 userdata Location field lat int32_t read write -900000000 900000000
 userdata Location field lng int32_t read write -1800000000 1800000000
+userdata Location field alt int32_t read write -INT32_MAX INT32_MAX
```
The max should be LOCATION_ALT_MAX_M, the minimum should probably be -((1 << 24) - 1) since we periodically need to be able to serialize this to 24 bit for missions.

rmackay9 replied:
How about this?  I think this will match [our existing check in AP_Mission](https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Mission/AP_Mission.cpp#L1003)

> userdata Location field alt int32_t read write (-LOCATION_ALT_MAX_M*100+1) (LOCATION_ALT_MAX_M*100-1)

rmackay9 commented on libraries/AP_Vehicle/AP_Vehicle.cpp:
```diff
@@ -95,6 +95,12 @@ void AP_Vehicle::init_vehicle()
 #endif
 }
 
+void AP_Vehicle::register_scheduler_delay_callback()
```
Yes, it's an unrelated change (and in a separate commit, "AP_Vehicle: formatting fixes").  I moved it to make it's implementation appear in the same order as it's declaration in the .h file.  Perhaps it's sligtly OCD of me but in Copter we've done an OK job of keeping things organised so I'd like to apply some of that organisation to AP_Vehicle because it looks like more Copter code is going to move there.

rmackay9 commented on libraries/AP_Scripting/examples/set-target-location.lua:
```diff
@@ -0,0 +1,53 @@
+            sent_target = vehicle:set_target_location(above_home)   -- set target above home
+          end
+        end
+      else
```
We could do that but it doesn't really make much difference I think.  With the else it spends a minimum of 1sec in Guided mode.  Without the else it still goes into Guided mode but for a very short period of time.

From my point of view, whenever we add a feature to scripting we should try and come up with an example script to show people how to use it but I think it's unlikely that anyone will really use this Lua version of RTL for anything more than a starting point to build something better.

rmackay9 commented:
I've fixed the one or two issues raised, rebased on master and re-tested this and it seems to be working.  I also tested the new altitude limit by attempting to set an RTL alt-above-home of 83000m (it failed) or 82000 (it succeeded).

rmackay9 commented:
rebased on master..

WickedShell commented:
Just need some plane and sub support! Thanks though.

rmackay9 commented:
Awesome, thanks!
